sys/drv/rtl8139.c

```diff
@@ -32,19 +39,55 @@ static int rtl_reset(rtl8139_state_t *state) {
   return 1;
 }
 
+static intr_filter_t rtl8139_intr(void *data) {
+  rtl8139_state_t *state = data;
+  /* TODO: add enqueue routines */
+  bus_write_2(state->regs, RL_ISR, RL_ISR_RX_OK);
+  return IF_STRAY;
+}
+
 static int rtl8139_attach(device_t *dev) {
   rtl8139_state_t *state = dev->state;
+  int err = 0;
+
+  pci_enable_busmaster(dev);
 
   state->regs = device_take_memory(dev, 1, RF_ACTIVE);
-  if (state->regs == NULL)
+  state->irq_res = device_take_irq(dev, 0, RF_ACTIVE);
+  if (!state->regs || !state->irq_res) {
+    klog("Failed to init resources!");
     return ENXIO;
+  }
+  bus_intr_setup(dev, state->irq_res, rtl8139_intr, NULL, state, "RTL8139");
```

Comment:
You should call `bus_intr_teardown` if you return from `attach` with value `!= 0`.

Reply:
is current solution acceptable?

Reply:
No.